docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,622 @@
+ *
+ * // Using custom configuration values to set max upload speed to avoid
+ * // saturating the network interface.
+ * S3TransferManager tm = S3TransferManager.builder()
+ *         .configuration(TransferManagerConfiguration.builder()
+ *             .maximumUploadBytesSecond(32 * 1024 * 1024) // 32 MiB
+ *             .build()
+ *         .build();
+ * }
+ * </pre>
+ */
+public interface S3TransferManager {
+    /**
+     * DownloadObject an object identified by the bucket and key from S3 to the
+     * given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * DownloadObject myFileDownload = tm.downloadObject(BUCKET, KEY, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload().completionFuture().join();
+     * }
+     * </pre>
+     *
+     */
+    default DownloadObject downloadObject(String bucket, String key, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .getObjectRequest(GetObjectRequest.builder()
+                                .bucket(bucket)
+                                .key(key)
+                                .build())
+                        .build(),
+                file);
+    }
+
+    /**
+     * DownloadObject an object using an S3 presigned URL to the given file.
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate transfer
+     * DownloadObject myFileDownload = tm.downloadObject(myPresignedUrl, Paths.get("/tmp/myFile.txt");
+     * // Wait for transfer to complete
+     * myFileDownload()completionFuture().join();
+     * }
+     * </pre>
+     */
+    default DownloadObject downloadObject(URL presignedUrl, Path file) {
+        return downloadObject(DownloadObjectRequest.builder()
+                        .presignedUrl(presignedUrl)
+                        .build(),
+                file);
+    }
+    /**
+     * DownloadObject an object in S3 to the given file.
+     *
+     * <p>
+     * <b>Usage Example:</b>
+     * <pre>
+     * {@code
+     * // Initiate the transfer
+     * DownloadObject myDownload = tm.downloadObject(DownloadObjectRequest.builder()
+     *         .getObjectRequest(GetObjectRequest.builder()
+     *                 .bucket(BUCKET)
+     *                 .key(KEY)
+     *                 .build())
+     *          // Set the known length of the object to avoid a HeadObject call
+     *         .hints(DownloadHints.builder()
+     *                 .objectLength(1024 * 1024 * 5)
+     *                 .build())
+     *         .build(),
+     *         Paths.get("/tmp/myFile.txt"));
+     * // Wait for the transfer to complete
+     * myDownload.completionFuture().join();
+     * }
+     * </pre>
+     */
+    default DownloadObject downloadObject(DownloadObjectRequest request, Path file) {
+        return downloadObject(request, ObjectPartTransformerCreator.forFile(file));
```

Comment:
Are we going to use parts or ranged Gets for downloads? I vote for ranged Gets as it can support all objects even if they are not uploaded in parts.

Reply:
Yes, the current plan is to use range gets for all downloads with a possible setting to allow parallel downloads only for multipart objects (like in V1) if there's a benefit to that.

Reply:
Even if using ranged gets we should still download on the part boundaries and S3 should have an API for retrieving the ranges. This should provide significant benefit both to the client and S3 itself.

Reply:
Yup, I think we would prefer to download on part boundaries if the object was uploaded multipart.